### PR TITLE
[Bug] Fix singleton class lookup bug when singleton class is replaced on the instance

### PR DESCRIPTION
**Pull request by itarato**
`IO#reopen` is one of those few examples (maybe the only one) where the caller object's singleton class is replaced:

https://github.com/oracle/truffleruby/blob/master/src/main/ruby/truffleruby/core/io.rb#L1901

This means, methods defined only on an IO object's singleton class won't be available after calling `#reopen` on it.

However, currently in TruffleRuby fetching the singleton class for an object is cached against the object - which makes an assumption that the same object must have the same singleton class during its lifetime. As the beginning states, this is not the case for IO objects currently.

An example where this occurs on production: ActiveSupport's test runner contains an IO capture mechanism: https://github.com/rails/rails/blob/0e99d0893b0e98c626a3c7d8972eea22d29c9d25/activesupport/lib/active_support/testing/stream.rb#L23-L38 This effectively erases `STDOUT`'s singleton class. Any extension on `STDOUT`'s singleton class, eg with [this public gem](https://github.com/Shopify/cli-ui/blob/f6294ff1864dc3279fa7ca08ec79d17b338fe1f5/lib/cli/ui/stdout_router.rb#L402-L410) could cause a lookup of the old singleton class function.

The actual break on production was a bit more nuanced:
- the linked gem re-aliased the `IO#write` method with a new proxy method, both on the singleton class (to enhance the functionality of `STDOUT.write`).
- a test was leveraging `ActiveSupport`'s `#capture` mechanism, which deleted the singleton class
- a follow up `STDOUT.write` was - due to the bug - directed to the old metaclass' `.write` method, which failed to call the old alias.

To sum up the bug's effect, after replacing a singleton class:
- singleton methods and aliases can be "mixed up" causing method-not-find
- new singleton methods/aliases not being found

### CODE REVIEW

File: src/main/java/org/truffleruby/language/objects/MetaClassNode.java [hunk 1]
```diff
@@ -42,7 +42,11 @@ public final RubyClass executeCached(Object value) {
     public abstract RubyClass execute(Node node, Object value);
 
     @Specialization(
-            guards = { "isSingleContext()", "object == cachedObject", "metaClass.isSingleton" },
+            guards = {
+                    "isSingleContext()",
+                    "object == cachedObject",
+                    "metaClass.isSingleton",
+                    "metaClass == object.getMetaClass()" },
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/objects/MetaClassNode.java:49]: This adds a runtime check and makes this specialization not really worth it anymore. Instead, we could guard on the cached object.getMetaClass() being not <= IO, or more simply `!(cachedObject instanceof RubyIO)`, that would then only be tested on the slow path and not the fast path.

+ change the vm_set_class primitive to only accepts RubyIO objects

Comment (itarato) [reply src/main/java/org/truffleruby/language/objects/MetaClassNode.java]: Done.

File: spec/ruby/language/method_spec.rb [hunk 2]
```diff
@@ -176,6 +176,22 @@ def m(x)
       m('a', &nil).should == [false]
     end
   end
+
+  context "with a replaced singleton class" do
```
Comment (eregon) [inline spec/ruby/language/method_spec.rb:180]: language is not a great place for this, especially it shouldn't `require "socket"`, I'll move this to IO#reopen, it's really an edge case caused by IO#reopen

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,7 @@ New features:
 Bug fixes:
 
 * Fix `Dir.glob` returning blank string entry with leading `**/` in glob and `base:` argument (@rwstauner).
+* Fixing singleton class lookup after an object's singleton class has been replaced (@itarato).
```
Comment (eregon) [inline CHANGELOG.md:9]:
```suggestion
* Fix class lookup after an object's class has been replaced by IO#reopen (@itarato).
```
I'll do it
